### PR TITLE
NewPM/AMDGPU: Port AMDGPUPerfHintAnalysis to new pass manager

### PR DESCRIPTION
**Pull request by arsenm**
This was much more difficult than I anticipated. The pass is
not in a good state, with poor test coverage. The legacy PM
does seem to be relying on maintaining the map state between
different SCCs, which seems bad. The pass is going out of its
way to avoid putting the attributes it introduces onto non-callee
functions. If it just added them, we could use them directly
instead of relying on the map, I would think.

The NewPM path uses a ModulePass; I'm not sure if we should be
using CGSCC here but there seems to be some missing infrastructure
to support backend defined ones.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102645?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#102645** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102645?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#102644** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/102644?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/AMDGPUPassRegistry.def [hunk 1]
```diff
@@ -22,6 +22,7 @@ MODULE_PASS("amdgpu-lower-buffer-fat-pointers",
             AMDGPULowerBufferFatPointersPass(*this))
 MODULE_PASS("amdgpu-lower-ctor-dtor", AMDGPUCtorDtorLoweringPass())
 MODULE_PASS("amdgpu-lower-module-lds", AMDGPULowerModuleLDSPass(*this))
+MODULE_PASS("amdgpu-perf-hint", AMDGPUPerfHintAnalysisPass(*static_cast<const GCNTargetMachine *>(this)))
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/AMDGPUPassRegistry.def:25]: Exceeds 80 chars per line.

File: llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp [hunk 2]
```diff
@@ -413,18 +439,57 @@ bool AMDGPUPerfHintAnalysis::runOnSCC(CallGraphSCC &SCC) {
...
 
-  return AMDGPUPerfHint::isMemBound(FI->second);
+  SmallVector<Function *, 16> Worklist;
+  CG.buildRefSCCs();
+  for (LazyCallGraph::RefSCC &RC : CG.postorder_ref_sccs()) {
+    for (LazyCallGraph::SCC &SCC : RC) {
+      if (SCC.size() != 1)
+        continue;
+      Function &F = SCC.begin()->getFunction();
+      if (!F.isDeclaration() && !F.doesNotRecurse() && F.hasInternalLinkage())
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp:452]: Why is it limited to internal linkage?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp:452]: Copied from FunctionAttrs, but I think this meant to really be checking for interposable linkage (and I thought I dropped this before posting?)

**Review by shiltian — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp]: Actually this broke the test

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUPerfHintAnalysis.cpp]: Nevermind, I somehow posted the wrong version that has the reverse list which is wrong

**Comment by arsenm** [comment]
### Merge activity

* **Aug 11, 7:00 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/102645).

**Pull request merged**